apps/api-extractor/src/analyzer/AstReferenceResolver.ts

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as ts from 'typescript';
+import * as tsdoc from '@microsoft/tsdoc';
+
+import { AstSymbolTable, AstEntity } from './AstSymbolTable';
+import { AstDeclaration } from './AstDeclaration';
+import { WorkingPackage } from '../collector/WorkingPackage';
+import { AstModule } from './AstModule';
+import { AstImport } from './AstImport';
+
+/**
+ * This resolves a TSDoc declaration reference by walking the `AstSymbolTable` compiler state.
+ *
+ * @remarks
+ *
+ * This class is analogous to `ModelReferenceResolver` from the `@microsoft/api-extractor-model` project,
+ * which resolves declaration references by walking the hierarchy loaded from an .api.json file.
+ */
+export class AstReferenceResolver {
+  private readonly _astSymbolTable: AstSymbolTable;
+  private readonly _workingPackage: WorkingPackage;
+
+  public constructor(astSymbolTable: AstSymbolTable, workingPackage: WorkingPackage) {
+    this._astSymbolTable = astSymbolTable;
+    this._workingPackage = workingPackage;
+  }
+
+  public resolve(declarationReference: tsdoc.DocDeclarationReference): AstDeclaration | Error {
```

Comment:
Why return `Error` instead of just throwing?

Reply:
These errors never get thrown as exceptions.  Really they are just containers for substrings that get incorporated into an `ExceptionMessage` object that gets routed.  I guess it's a little confusing to use the `Error` class for this.  I was trying to avoid creating a new type, but maybe that would be more clear.

Reply:
Fixed